block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagRepresentationSyntax.scala

```diff
@@ -79,25 +65,100 @@ final class BlockDagRepresentationOps[F[_]: Sync](
+      implicit sync: Sync[F]
   ): F[Map[Validator, BlockHash]] =
     dag.invalidBlocks.map { invalidBlocks =>
       latestMessagesHashes.filter {
         case (_, blockHash) => invalidBlocks.map(_.blockHash).contains(blockHash)
       }
     }
 
-  def invalidBlocksMap: F[Map[BlockHash, Validator]] =
+  def invalidBlocksMap(implicit sync: Sync[F]): F[Map[BlockHash, Validator]] =
     for {
       ib <- dag.invalidBlocks
       r  = ib.map(block => (block.blockHash, block.sender)).toMap
     } yield r
+
+  def selfJustificationChain(h: BlockHash)(implicit sync: Sync[F]): Stream[F, Justification] =
+    Stream.unfoldEval(h)(
+      message =>
+        lookupUnsafe(message)
+          .map { v =>
+            v.justifications.find(_.validator == v.sender)
+          }
+          .map(_.map(next => (next, next.latestBlockHash)))
+    )
+
+  def selfJustification(h: BlockHash)(implicit sync: Sync[F]): F[Option[Justification]] =
+    selfJustificationChain(h).head.compile.last
+
+  def mainParentChain(h: BlockHash, stopAtHeight: Long = 0)(
+      implicit sync: Sync[F]
+  ): Stream[F, BlockHash] =
+    Stream.unfoldEval(h) { message =>
+      lookupUnsafe(message).map(
+        meta =>
+          if (meta.blockNum <= stopAtHeight)
+            none[(BlockHash, BlockHash)]
+          else
+            meta.parents.headOption.map(v => (v, v))
+      )
+    }
+
+  def isInMainChain(ancestor: BlockHash, descendant: BlockHash)(
+      implicit sync: Sync[F]
+  ): F[Boolean] = {
+    val result = for {
+      aHeight <- OptionT(dag.lookup(ancestor).map(_.map(_.blockNum)))
+      is = mainParentChain(descendant, aHeight)
+        .filter(_ == ancestor)
+        .head
+        .compile
+        .last
+        .map(_.isDefined)
+      r <- OptionT.liftF(is)
+    } yield r
+    (descendant == ancestor).pure ||^ result.getOrElse(false)
+  }
+
+  def parentsUnsafe(item: BlockHash)(implicit sync: Sync[F]): F[List[BlockHash]] = {
+    def errMsg = s"Parents lookup failed: DAG is missing ${item.show}"
+    dag.lookup(item).map(_.map(v => v.parents)) >>= (_.liftTo(BlockDagInconsistencyError(errMsg)))
+  }
+
+  def nonFinalizedBlocks(implicit sync: Sync[F]): F[Set[BlockHash]] =
+    for {
+      tips <- latestMessages.map(_.values.map(_.blockHash).toList)
+      r <- Stream
+            .unfoldLoopEval(tips) { lvl =>
+              for {
+                out  <- lvl.filterA(dag.isFinalized(_).not)
+                next <- out.traverse(dag.lookup(_).map(_.map(_.parents))).map(_.flatten.flatten)
+              } yield (out, next.nonEmpty.guard[Option].as(next))
+            }
+            .flatMap(Stream.emits)
+            .compile
+            .to(Set)
+    } yield r
+
+  def descendants(blockHash: BlockHash)(implicit sync: Sync[F]): F[Set[BlockHash]] =
+    Stream
+      .unfoldLoopEval(List(blockHash)) { lvl =>
+        for {
+          out  <- lvl.traverse(dag.children).map(_.flatten.flatten)
+          next = out
```

Comment:
I wanted to be explicit here, because `unfoldLoopEval ` always wants out and next. This logic assumes that next = out. But in function above, e.g. this is not the case. So I just found it more clear.